Micro16C.Web/src/Micro16C.Web.Client/Main.fs

```diff
@@ -0,0 +1,163 @@
+        |> PassManager.queueTransform Passes.deadCodeEliminationPass
+        |> PassManager.queueTransform Passes.instructionSimplifyPass
+        |> PassManager.queueTransform Passes.simplifyCFGPass
+        |> PassManager.queueTransform Legalize.legalizeInstructionsPass
+        |> PassManager.queueTransform Passes.constantPropagationPass
+        |> PassManager.queueTransform Passes.deadCodeEliminationPass
+        |> PassManager.queueTransform Passes.instructionSimplifyPass
+        |> PassManager.queueTransform Passes.simplifyCFGPass
+        |> PassManager.queueTransform Passes.deadBitEliminationPass
+        |> PassManager.queueTransform Passes.jumpThreadingPass
+        |> PassManager.queueTransform Passes.constantPropagationPass
+        |> PassManager.queueTransform Passes.deadCodeEliminationPass
+        |> PassManager.queueTransform Passes.instructionSimplifyPass
+        |> PassManager.queueTransform Passes.simplifyCFGPass
+        |> PassManager.queueTransform Passes.deadBitEliminationPass
+        |> PassManager.queueTransform (debugModulePasses "After optimizations:")
+        |> PassManager.queueTransform Legalize.legalizeConstantsPass
+        |> PassManager.queueTransform Legalize.breakPhiCriticalEdgesPass
+        |> PassManager.queueTransform Passes.reorderBasicBlocksPass
+        |> PassManager.queueTransform (debugModulePasses "End of IR:")
+        |> PassManager.queueTransform GenAssembly.genAssemblyPass
+        |> PassManager.queueTransform GenAssembly.removeUnusedLabelsPass
+        |> PassManager.queueTransform GenAssembly.removeRedundantLabelsPass
+
+    Lex.tokenize text
+    |> Result.bind Parse.parse
+    |> Result.bind Sema.analyse
+    |> Result.map Codegen.codegen
+    |> Result.map (fun x -> passManager |> PassManager.run x)
+
+
+/// The Elmish application's model.
+type Model =
+    {
+        counter: int
+        error: string option
+        codeToCompile: string
+        generatedCode: string
+    }
+
+
+let initModel =
+    {
+        counter = 0
+        error = None
+        codeToCompile = "int r0 = R0;
+int r1 = R1;
+int mod;
+do
+{
+    mod = r1 % r0;
+    r1 = r0;
+    r0 = mod;
+}
+while(mod != 0);
+R2 = r1;"
+        generatedCode = ""
+    }
+
+
+/// The Elmish application's update messages.
+type Message =
+    | Increment
+    | Decrement
+    | SetCounter of int
+    | Error of exn
+    | ClearError
+    | CompileCode of string
+
+let update message model =
+    match message with
+
+    | Increment ->
+        { model with counter = model.counter + 1 }
+    | Decrement ->
+        { model with counter = model.counter - 1 }
+    | SetCounter value ->
+        { model with counter = value }
+
+    | Error exn ->
+        { model with error = Some exn.Message }
+    | ClearError ->
+        { model with error = None }
+    | CompileCode value ->
+        { model with codeToCompile = value; generatedCode = match compile value |> Result.map Assembly.asText with 
```

Comment:
You can use `Result.mapError` here

Reply:
Sorry for being an F# noob, but what's that? And how would I use it?

Reply:
It works basically just the same as `Result.map` but on the error value instead of the Ok value. Eg. given `r` of type `Result<T,E>`, `Result.mapError` takes as first parameter a function `E -> F` and as second parameter the result and returns a `Result<T,F>`. If the result passed in is an `Ok` it goes through unchanged, if it is an `Error` however, the `Error` value gets mapped using the provided function.

It's basically the F# version of https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err in Rust.

Reply:
I see, thanks for the explanation! Though, all I really wanted to do was to unwrap the result and either display the string or to display a string representation of the error. And for that, I haven't found solution other than `match`?

Reply:
Ahh yes, true my bad. We are mapping to two strings, not a Result<string,string>